core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -866,6 +1161,181 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
+struct MDThreadVectorRange<Rank<N, OuterDir, InnerDir>, TeamHandle> {
+  using rank_type        = int;
+  using boundary_type    = int;
+  using team_handle_type = TeamHandle;
+  using execution_space  = typename team_handle_type::execution_space;
+  using array_layout     = typename execution_space::array_layout;
+  static constexpr rank_type total_rank = Rank<N, OuterDir, InnerDir>::rank;
+  static constexpr Iterate iter         = OuterDir;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<array_layout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION MDThreadVectorRange(team_handle_type const& team_,
+                                             Args&&... args)
+      : team(team_), boundaries{static_cast<boundary_type>(args)...} {
+    static_assert(sizeof...(Args) == total_rank);
+  }
+
+  team_handle_type const& team;
+  boundary_type boundaries[total_rank];
+};
+
+template <typename TeamHandle, typename... Args>
+MDThreadVectorRange(TeamHandle const&, Args&&...)
+    ->MDThreadVectorRange<Rank<sizeof...(Args), Iterate::Default>, TeamHandle>;
+
+template <typename Rank, typename TeamHandle>
+struct MDTeamVectorRange;
+
+template <unsigned N, Iterate OuterDir, Iterate InnerDir, typename TeamHandle>
+struct MDTeamVectorRange<Rank<N, OuterDir, InnerDir>, TeamHandle> {
+  using rank_type        = int;
+  using boundary_type    = int;
+  using team_handle_type = TeamHandle;
+  using execution_space  = typename team_handle_type::execution_space;
+  using array_layout     = typename execution_space::array_layout;
+  static constexpr rank_type total_rank = Rank<N, OuterDir, InnerDir>::rank;
+  static constexpr Iterate iter         = OuterDir;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<array_layout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION MDTeamVectorRange(team_handle_type const& team_,
+                                           Args&&... args)
+      : team(team_), boundaries{static_cast<boundary_type>(args)...} {
+    static_assert(sizeof...(Args) == total_rank);
+  }
+
+  team_handle_type const& team;
+  boundary_type boundaries[total_rank];
+};
+
+template <typename TeamHandle, typename... Args>
+MDTeamVectorRange(TeamHandle const&, Args&&...)
+    ->MDTeamVectorRange<Rank<sizeof...(Args), Iterate::Default>, TeamHandle>;
+
+template <typename Rank, typename TeamHandle, typename Lambda,
+          typename ReducerValueType>
+KOKKOS_INLINE_FUNCTION void parallel_reduce(
+    MDTeamThreadRange<Rank, TeamHandle> const& policy, Lambda const& lambda,
+    ReducerValueType& val) {
+  Impl::MDTeamRangeMode<Impl::MDTeamRangeEndRank::NotEndRank,
+                        Impl::MDTeamRangeParThread::ParThread,
+                        Impl::MDTeamRangeParVector::NotParVector>
+      rangeMode;
```

Comment:
I was looking into this, because it sounded like a plausible change to make at first.
But the "mode" is dictated by the policy only at its initial state. After that, it is updated based on the current rank that the recursion is operating on. So it is not set just based on the policy type. This is how "mode" is updated in each nested level:
```
  using RangeMode = TeamMDRangeMode<
      (iter == Iterate::Right)
          ? static_cast<TeamMDRangeEndRank>(current_rank == total_rank)
          : static_cast<TeamMDRangeEndRank>(current_rank == -1),
      static_cast<TeamMDRangeParThread>(current_rank == par_thread_rank),
      static_cast<TeamMDRangeParVector>(current_rank == par_vector_rank)>;
 ```
I did have another implementation that didn't use a tag class like this, but I don't believe that it was much better than the current implementation.

Reply:
Handled in [54df360](https://github.com/kokkos/kokkos/pull/5238/commits/54df360c6ee32596442f3ad544cd7648e26a6e52). Removed these explicit modes from the interface. They are now initially deduced from the user input policy, but a form of a "mode" is still kept track of in impl.